### ISSUE TITLE
[css-mediaqueries] Proposal: "prefers-reduced-data"

### ISSUE THREAD

**Issue by Malvoz** [w3c/csswg-drafts#2370]
Would CSSWG consider this proposal of <s>detecting</s> exposing if the [`Save-Data`](http://httpwg.org/http-extensions/client-hints.html#save-data) request header from the [Client Hint](https://github.com/httpwg/http-extensions#client-hints) HTTP extension is on/off to allow for usage with CSS?

### The `Save-Data` HTTP request header:
> ...a signal indicating explicit user opt-in into a reduced data usage mode on the client, and when communicated to origins allows them to deliver alternate content honoring such preference - e.g. smaller image and video resources, alternate markup...<br>
_[- http://httpwg.org/http-extensions/client-hints.html#save-data](http://httpwg.org/http-extensions/client-hints.html#save-data)_

### The 'prefers-reduced-data' feature:
<table>
  <tr>
    <th>Name:</th>
    <th>For:</th>
     <th>Value:</th>
<th>Type:</th>
  </tr>
  <tr>
    <td><code>prefers-reduced-data</code></td>
    <td><code>@media</code></td>
    <td><code>no-preference | reduce</code></td>
<td><code>discrete</code></td>
  </tr>
</table>

### Example usage:
```
.image {
  background-image: url("image/heavy.jpg");
}

@media (prefers-reduced-data: reduce) { 
  /* Save-Data: On */
  .image {
    background-image: url("image/light.webp");
  }
}
```

Relevant articles, including use cases:
- [https://developers.google.com/web/updates/2016/02/save-data](https://developers.google.com/web/updates/2016/02/save-data)
- [https://www.ctrl.blog/entry/http-save-data](https://www.ctrl.blog/entry/http-save-data)
- [https://css-tricks.com/help-users-save-data/](https://css-tricks.com/help-users-save-data/)

**Comment by frivoal** [comment]
Well, this would not **detect** the HTTP header, since that is sent from the client to the server, and this runs on the client, but we could have this and trigger it on the same conditions that cause the client to send that header.

**Comment by frivoal** [comment]
The CSS-WG has considered this, and thinks this sounds generally reasonable, but we would like to hear more about what this would be used for. The cases that were brought up during our meeting all already had solutions.

Valid use cases for this that aren't already addressed by any existing thing would be a good justification to add this, so we're looking for examples of that.

**Comment by Malvoz** [comment]
Adding example to use cases:
`@import "this-awesome-large-theme.css" not (prefers-reduced-data: reduce);`.

This and the initial use cases could potentially all be done using scripting, as would many existing `@media` queries too. Will just have to wait and see how other developers recieves this proposal. Thanks for considering this.

**Comment by Malvoz** [comment]
@tabatkins 

> What's this-awesome-large-theme.css? Just a really big CSS file? 

In this example, yes.

> So you'd presumably send a much smaller, more basic CSS file instead for the reduce case?

Yes. Or just keep the "[app shell](https://developers.google.com/web/fundamentals/architecture/app-shell)" CSS in that main file and simply not `@import` any extra fancy styles (and fonts, images etc.) when user prefers reduced data.

**Comment by tabatkins** [comment]
I think that's a reasonable use-case that *can't* be addressed by any other mechanism. @frivoal, what do you think?

**Comment by frivoal** [comment]
It seems sane to me. As for implementer interest, that's not for me to say. MQ5 is still a very early draft, so I don't think the implementer interest bar has to be very high, but I'd like to see at least a nod in favor of this feature before adding it. I guess you (@tabatkins) counts as one.

**Comment by addyosmani** [comment]
@yoavweiss and I were about to propose a `prefers-reduced-data` media query and saw that this issue had already been filed. @frivoal, I guess beyond Tab you are still waiting on interest from one other implementor?

**Comment by frivoal** [comment]
> I guess beyond Tab you are still waiting on interest from one other implementor?

Yes.

**Comment by frivoal** [comment]
Then again, it's not strictly gated on multi-vendor interest. Clear interest from Google counts for more than just Tab feeling OK with it.

I would still appreciate a detailed list of uses cases, and why existing things aren't good enough to solve them. That would help make sure the feature actually solves the problem it is meant to solve. https://github.com/w3c/csswg-drafts/issues/2370#issuecomment-415482857 is a start, but it'd be nice to see more detail or more cases.

**Comment by ryantownsend** [comment]
@frivoal I was chatting about this with @yoavweiss at performance.now() conf and he sent me just this thread.

I have a use-case for `prefers-reduced-data` in that I’m avoiding loading webfonts and some iconography (instead leaving textual labels) based on this flag.

Currently, I have an inline JS script in my `<head>` tag which adds a CSS class to the document – it’s only a small script so it’s pretty performant but it would be far cleaner if I didn’t have to rely on that. My workaround also breaks in cases where someone is blocking JS to also double-down on data savings (given the minimal adoption of `Save-Data`).

I feel like this would be useful for websites without real-time image transformation CDNs or ones which do not respond to Client Hint headers – so it would be potentially useful if this media query was available to `<source>`.

Hope this is helpful, please do reach out if I can provide any further context.

**Comment by Schepp** [comment]
I find the use cases Ryan outlines very interesting. On top of `<source>` elements this could also help with controlling resource hints.

**Comment by mrego** [comment]
@frivoal shouldn't this be rediscussed withing the working group, it seems people was positive more than a year ago, we got some interesting use cases, but then the text was added to the spec without asking for review to the rest of the group. WDTY?

**Comment by mrego** [comment]
Also regarding the example, wouldn't be better to use the webfont example instead of the image thing that can be confused with use cases already covered by responsive images?

**Comment by yoavweiss** [comment]
Regarding a privacy section for SaveData in general, there's https://wicg.github.io/savedata/#privacy-considerations
If insufficient, please open issues on that draft, and just point the relevant CSS spec to it.

Also, is there a privacy section for the `prefers-reduced-motion` MQ?
